extensions/amp-animation/0.1/web-animations.js

```diff
@@ -84,18 +86,194 @@ const SERVICE_PROPS = {
 
 /**
  */
-export class WebAnimationRunner {
+export class AnimationRunner {
 
   /**
    * @param {!Array<!InternalWebAnimationRequestDef>} requests
    */
   constructor(requests) {
-    /** @const @private */
+    /** @const @protected */
     this.requests_ = requests;
+  }
+
+  /**
+   * @return {!WebAnimationPlayState}
+   */
+  getPlayState() {
+  }
+
+  /**
+   * @param {function(!WebAnimationPlayState)} unusedHandler
+   * @return {!UnlistenDef}
+   */
+  onPlayStateChanged(unusedHandler) {
+  }
+
+  /**
+  * Initializes the players but does not change the state.
+   */
+  init() {
+  }
+
+  /**
+   * Initializes the players if not already initialized,
+   * and starts playing the animations.
+   */
+  start() {
+  }
+
+  /**
+   */
+  pause() {
+  }
+
+  /**
+   */
+  resume() {
+  }
+
+  /**
+   */
+  reverse() {
+  }
+
+  /**
+   * @param {time} unusedTime
+   */
+  seekTo(unusedTime) {
+  }
+
+  /**
+   * Seeks to a relative position within the animation timeline given a
+   * percentage (0 to 1 number).
+   * @param {number} unusedPercent between 0 and 1
+   */
+  seekToPercent(unusedPercent) {
+  }
+
+  /**
+   */
+  finish() {
+  }
+
+  /**
+   */
+  cancel() {
+  }
+
+  /**
+   * @param {!WebAnimationPlayState} unusedPlayState
+   * @private
+   */
+  setPlayState_(unusedPlayState) {
+  }
+}
+
+/**
+ */
+export class AnimationWorkletRunner extends AnimationRunner {
 
-    /** @private {?Array<!Animation>} */
+  /**
+   * @param {!Window} win
+   * @param {!Array<!InternalWebAnimationRequestDef>} requests
+   */
+  constructor(win, requests) {
+    super(requests);
+
+    /** @const @private */
+    this.win_ = win;
+
+    /** @protected {?Array<!WorkletAnimation>} */
     this.players_ = null;
+  }
+
+  /**
+   * @return {string}
+   */
+  createCodeBlob_() {
+
+    return `
+    registerAnimator('anim${++animIdCounter}', class {
+      animate(currentTime, effect) {
+        // TODO: Do some work with \`viewport-margins\` here.
+        if (currentTime == NaN) {
+          return;
+        }
+        effect.localTime = currentTime;
+      }
+    });
+    `;
+  }
+
+  /**
+  * @override
+  * Initializes the players but does not change the state.
+   */
+  init() {
+    this.requests_.map(request => {
+      // Apply vars.
+      if (request.vars) {
+        setStyles(request.target,
+            assertDoesNotContainDisplay(request.vars));
+      }
+      // TODO(nainar): This switches all animations to AnimationWorklet.
+      // Limit only to Scroll based animations for now.
+      CSS.animationWorklet.addModule(
+          URL.createObjectURL(new Blob([this.createCodeBlob_()],
+              {type: 'text/javascript'}))).then(() => {
+        const scrollSource = this.win_.document./*OK*/scrollingElement;
```

Comment:
This differs depending on the viewport (e.g. iOS in embed mode has a completely different scroller which is not body), must use viewport service: `viewport.getScrollingElement()`

Reply:
Done. `Viewport` didn't have a public `getScrollingElement()` so I added one. It does have access to the `scrollingElement` through the `ViewportBinding` though.